### PR TITLE
[Cosmos] Adds Configurable Max Retry Attempts for Throttled (HTTP 429) Requests

### LINKED ISSUE #4543: [Cosmos] Support configurable max retry attempts/wait time for rate-limited (HTTP 429) requests
# Support configurable max retry attempts/wait time for rate-limited (HTTP 429) requests

## Summary

The Rust Cosmos DB driver and SDK should let callers configure how many times a throttled (HTTP 429, rate-limited) request is retried, and how long retrying may take in aggregate — matching the .NET SDK's [`CosmosClientOptions.MaxRetryAttemptsOnRateLimitedRequests`](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/main/Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs#L456) and `MaxRetryWaitTimeOnRateLimitedRequests`.

Today the throttle retry budget is hard-coded (9 attempts / 30s), with no way to tune it per-request or per-client.

## Motivation

- **.NET parity** — applications migrating from the .NET SDK expect to control 429 retry behavior.
- **Latency-sensitive workloads** — some callers want to fail fast (`0` retries) and handle 429s themselves rather than absorbing up to 30s of internal backoff.
- **Bulk/ingestion workloads** — others want to extend the retry budget.

### Driver (`azure_data_cosmos_driver`) — `OperationOptions`

| Field | Type | Env var | Default | Notes |
| --- | --- | --- | --- | --- |
| `max_throttle_retry_count` | `Option<u32>` | `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT` | `9` | `0` disables throttle retries (first 429 surfaces to caller) |
| `max_throttle_retry_wait_time` | `Option<Duration>` | — (none) | `30s` | No env var — `Duration` is not `FromStr`-parseable |

Resolves through the standard operation → account → runtime → env layering and is threaded into the transport-level throttle retry loop.

### SDK (`azure_data_cosmos`)

- Per-request: settable via the re-exported `OperationOptions` / `OperationOptionsBuilder`.
- Client-wide: new `CosmosClientBuilder` methods
  - `with_max_retry_attempts_on_throttled_requests(u32)`
  - `with_max_retry_wait_time_on_throttled_requests(Duration)`

  forwarded to the driver's runtime layer at `build()` time.

## Behavior notes

- `max_throttle_retry_count = 0` → the first 429 propagates with no retry.
- `max_throttle_retry_wait_time` caps cumulative backoff; once the next delay would exceed the budget, the 429 propagates.
- The service `x-ms-retry-after-ms` header is still honored (and capped per-retry) within these limits.
- This is the **transport-level 429 retry loop** — distinct from the cross-region failover classifier that treats `429/3092 SystemResourceUnavailable` as a failover signal.

## Acceptance criteria

- [ ] Driver `OperationOptions` exposes both fields with documented defaults and env var.
- [ ] Throttle retry loop honors configured count + wait time; `0` disables retries.
- [ ] SDK exposes per-request override and client-wide builder methods.
- [ ] Unit tests: disable (`0`), custom count cap, custom wait-time cap, option round-trip + env-var parsing, SDK builder wiring.
- [ ] Docs (`ConfigurationOptions.md`) and CHANGELOG entries for both crates.

### PR DESCRIPTION
**Pull request by kundadebdatta**
# Add configurable max retry attempts/wait time for throttled (HTTP 429) requests

## Summary

Adds .NET-parity configuration for retrying rate-limited (HTTP 429) requests in the Cosmos DB Rust driver and SDK. Previously the throttle retry budget was hard-coded (9 attempts / 30s cumulative wait) with no way to tune it. This PR makes both limits configurable per-request and client-wide, mirroring the .NET SDK's [`MaxRetryAttemptsOnRateLimitedRequests`](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/main/Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs#L456) / `MaxRetryWaitTimeOnRateLimitedRequests`.

## Motivation

- **.NET parity** — callers migrating from the .NET SDK expect to control 429 retry behavior.
- **Fail-fast workloads** — latency-sensitive callers can set `0` retries and handle 429s themselves instead of absorbing up to 30s of internal backoff.
- **Bulk/ingestion workloads** — others can extend the retry budget.

### Driver (`azure_data_cosmos_driver`)

- `OperationOptions` gains two fields:

  | Field | Type | Env var | Default | Notes |
  | --- | --- | --- | --- | --- |
  | `max_throttle_retry_count` | `Option<u32>` | `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT` | `9` | `0` disables throttle retries (first 429 propagates) |
  | `max_throttle_retry_wait_time` | `Option<Duration>` | — (none) | `30s` | No env var — `Duration` isn't `FromStr`-parseable, following the `endpoint_unavailability_ttl` precedent |

- New `ThrottleRetryState::with_limits(max_attempts, max_wait_time)` constructor; production now builds throttle state from the resolved limits. The legacy `new()` (defaults) is now `#[cfg(test)]`.
- `TransportPipelineContext` carries `max_throttle_attempts` / `max_throttle_wait_time`; the operation pipeline resolves them through the standard **operation → account → runtime → env** layering and passes them into the transport-level 429 retry loop.

### SDK (`azure_data_cosmos`)

- Per-request: settable via the already re-exported `OperationOptions` / `OperationOptionsBuilder`.
- Client-wide: new `CosmosClientBuilder` methods
  - `with_max_retry_attempts_on_throttled_requests(u32)`
  - `with_max_retry_wait_time_on_throttled_requests(Duration)`
- `CosmosClientBuilder::build()` now forwards client-level operation options into the driver's runtime layer (previously `self.options.operation` was dropped and only PPCB was forwarded). PPCB default-forcing behavior is preserved exactly.

## Behavior notes

- `max_throttle_retry_count = 0` → the first 429 propagates with no retry.
- `max_throttle_retry_wait_time` caps cumulative backoff; once the next delay would exceed the budget, the 429 propagates.
- The service `x-ms-retry-after-ms` header is still honored (and capped per-retry) within these limits.
- This is the **transport-level 429 retry loop** — intentionally distinct from the cross-region failover classifier that treats `429/3092 SystemResourceUnavailable` as a failover signal. The two paths remain separate.

## Tests

- Driver transport (`transport_pipeline.rs`): `max_attempts = 0` disables; custom count cap honored; custom wait-time cap honored.
- Driver `components.rs`: `with_limits` overrides; `new()` matches `with_limits` defaults.
- Driver `operation_options.rs`: builder round-trip and `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT` env-var parsing (wait-time has no env var).
- SDK `cosmos_client_builder.rs`: builder setters populate operation options; `0` round-trips.
- Results: driver lib **1611 passed**, SDK lib **90 passed**.

## Docs & changelog

- `azure_data_cosmos/docs/ConfigurationOptions.md` — rows for both new options.
- CHANGELOG entries in both `azure_data_cosmos` and `azure_data_cosmos_driver`.

## Validation

`cargo fmt`, `build`, `clippy`, `doc`, and `test` all clean on both crates. (Pre-existing, unrelated failures remain only under `--all-features` due to missing OpenSSL on the dev machine, and in some integration tests/examples requiring the `key_auth`/hedging features — none touched by this PR.)

## Files changed (8 files, +260 / −10)

| File | +/− |
| --- | --- |
| `azure_data_cosmos/CHANGELOG.md` | +2 |
| `azure_data_cosmos/docs/ConfigurationOptions.md` | +2 |
| `azure_data_cosmos/src/clients/cosmos_client_builder.rs` | +80 |
| `azure_data_cosmos_driver/CHANGELOG.md` | +2 |
| `azure_data_cosmos_driver/src/driver/pipeline/components.rs` | +43 |
| `azure_data_cosmos_driver/src/driver/pipeline/operation_pipeline.rs` | +20 |
| `azure_data_cosmos_driver/src/driver/transport/transport_pipeline.rs` | +88 |
| `azure_data_cosmos_driver/src/options/operation_options.rs` | +33 |

## Breaking changes

None. Both fields are additive `Option<T>`s on a `#[non_exhaustive]` struct; defaults preserve existing behavior (9 / 30s).

### CODE REVIEW

**Review by analogrelay — COMMENTED** [review]
Looks pretty good to me. Can we get a test in using the fault-injection stuff and the emulator? Other than that I think it's good to go.

File: sdk/cosmos/azure_data_cosmos/docs/ConfigurationOptions.md [hunk 1]
```diff
@@ -157,6 +157,8 @@ pub struct OperationOptions { /* fields below */ }
...
 | `content_response_on_write` | `Option<bool>` | `AZURE_COSMOS_CONTENT_RESPONSE_ON_WRITE` | Whether write operations return the resource body in the response. Only applicable to write operations; ignored by reads and queries. Cascades from runtime → account → operation, matching .NET/Java/Go behavior. |
+| `max_throttle_retry_count` | `Option<u32>` | `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT` | Maximum number of retries when a request is throttled (HTTP 429, rate-limited) before the error surfaces to the caller. Analogous to .NET's `MaxRetryAttemptsOnRateLimitedRequests`. Defaults to `9` when unset; `0` disables throttle retries entirely. **Scope**: applies per transport-pipeline invocation, not per logical operation — an operation that fans out across regions (failover, hedging) gets a fresh budget per leg. Use `end_to_end_latency_policy` to bound total per-operation time. Settable client-wide via [`CosmosClientBuilder::with_max_retry_attempts_on_throttled_requests`]. |
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure_data_cosmos/docs/ConfigurationOptions.md]: My 2 cents - have one struct for ThrottlingRetryOptions insteadof the two separate settings

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure_data_cosmos/docs/ConfigurationOptions.md]: Also - do we really want to keep 30 seconds - seems a lot - now is the time to consider any changes to the defaults?

Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/docs/ConfigurationOptions.md]: Yeah - I kept it as `30s` to match with Java and .NET defaults. However, if we would like to reduce it, do you think `15 sec` would be a good to start with ? I think we can easily fit 9 retry attempts during this timeframe.

Given that chaning default is a small task, and requires some follow up with the PMs, we can defer the decision until the GA timeline. Discussed offline with @FabianMeiswinkel .

Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/docs/ConfigurationOptions.md]: On the refactoring of `ThrottlingRetryOptions` I agree and this will make things much simpler. This is now addressed with the latest commit.

**Comment by kundadebdatta** [comment]
## PR Deep Review (Round 3) → Summary

Reviewed the delta since my prior round (head `b6b29e4d2` → `df5aaedab`). The refactor to the nested `ThrottlingRetryOptions` group is well-executed and integrates the fault-injection coverage @analogrelay asked for. **No new 🔴 blockers**  the prior 🔴 forced-final-retry gap is fixed and `execute_transport_pipeline_with_zero_max_attempts_does_not_retry_429` pins it as a regression guard. 1615 driver tests pass at HEAD.

**6 findings posted** (3 🟡 Recommendations, 2 🟢 Suggestions, 1 💬 Observation):

| # | Severity | Category | TL;DR |
|---|---|---|---|
| 1 | 🟡 | Correctness: Off-by-one | `max_retry_count = N` produces `N + 2` attempts on the wire (1 initial + N retries + 1 forced-final). Doc says "N retries"; reality is N+1 retries. .NET-parity gap (`MaxRetryAttemptsOnRateLimitedRequests = 9` yields 10 attempts in .NET vs 11 in Rust). |
| 2 | 🟡 | Maintainability: Macro Recursion | `runtime.rs:806` requires manual `ThrottlingRetryOptions::from_env()` because the `CosmosOptions` derive doesn't recurse into `#[option(nested)]` fields. Future nested option groups will silently lose env support. |
| 3 | 🟡 | Open Thread: Defaults | @FabianMeiswinkel's open question (30s vs 15s default for `max_retry_wait_time`) is unresolved. Worth deciding now  once shipped, lowering the default is behavior-breaking for workloads that depended on absorbing up to 30s of throttle backoff. |
| 4 | 🟢 | Maintainability: Latent Footgun | `cosmos_client_builder::build()` still unconditionally clobbers `per_partition_circuit_breaker_enabled` (carry-over from my prior round's #6). |
| 5 | 🟢 | Maintainability: Test Coverage | No unit test exercises the nested `ThrottlingRetryOptions` layered-override case (operation-layer `count=0` over runtime-layer `count=9`). |
| 6 | 💬 | Net Observation | The PR is in significantly better shape than Round 2. Strong test coverage end-to-end via `fault_injection_429_honors_configurable_throttle_retry_count`; my prior fix preserved verbatim; CHANGELOGs updated correctly. |

The only finding I'd actually want resolved before merge is **#3** (the open default-value thread). #1 and #2 can be addressed as follow-ups if needed.

 I'm also tracking 🟢 #4 (PPCB clobber) from my prior round as still open. Not posting again to avoid dup-noise; flagging it inline.

---
<sub>⚠️ AI-generated review  may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs [hunk 2]
```diff
@@ -177,6 +193,54 @@ pub struct OperationOptions {
...
+#[non_exhaustive]
+pub struct ThrottlingRetryOptions {
+    /// Maximum number of retries when a request is throttled by the service
+    /// (HTTP 429, rate-limited).
+    ///
+    /// This is the analog of the .NET SDK's
+    /// `MaxRetryAttemptsOnRateLimitedRequests` (and Java's
+    /// `maxRetryAttemptsOnThrottledRequests`).
+    ///
+    /// **Default**: `9`. A value of `0` disables retrying throttled requests
```
Comment (kundadebdatta) [inline sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs:227]: **🟡 Recommendation** · Correctness: Off-by-one (— .NET parity)

**`max_retry_count = N` produces `N + 2` attempts on the wire, not `N + 1`**

The doc here says `Default: 9. A value of 0 disables retrying throttled requests (the first 429 is surfaced to the caller)` — a reasonable read is "9 retries = 10 attempts total". The author's own new test `execute_transport_pipeline_honors_configured_max_throttle_attempts` (and the integration test `fault_injection_429_honors_configurable_throttle_retry_count`) encode the actual formula:

`
ust
// 1 initial + N throttle retries + 1 forced-final-retry = N + 2.
let expected = max_throttle_attempts as usize + 2;
`

So:
- .NET `MaxRetryAttemptsOnRateLimitedRequests = 9` → 9 retries → **10** attempts
- Rust `max_retry_count = 9` (default) → 9 retries + 1 forced-final → **11** attempts

Off-by-one vs the ".NET parity" the PR description and CHANGELOG both lead with. Pick one fix:

**(a)** Document the forced-final-retry safety net in the user-facing field docs:

```suggestion
    /// Maximum number of retries when a request is throttled by the service
    /// (HTTP 429, rate-limited).
    ///
    /// This is the analog of the .NET SDK's
    /// `MaxRetryAttemptsOnRateLimitedRequests` (and Java's
    /// `maxRetryAttemptsOnThrottledRequests`).
    ///
    /// **Default**: `9`. A value of `0` disables retrying throttled requests
    /// (the first 429 is surfaced to the caller).
    ///
    /// **Note**: in addition to `N` configured retries, the driver performs
    /// one optional final retry attempt if the operation deadline still has
    /// remaining budget. So a non-zero `max_retry_count = N` produces up to
    /// `N + 2` requests on the wire (1 initial + N retries + 1 forced-final).
    /// `max_retry_count = 0` suppresses both regular retries *and* the
    /// forced-final retry.
```

**(b) (cleaner)** Cap total user-visible retries at `max_retry_count` by counting the forced-final-retry against the budget in `execute_transport_pipeline`: gate `forced_final_retry` on `attempt_count < max_attempts`, not just `max_attempts > 0`. Either way, the tests' `N + 2` formula and the user-facing doc should agree on what `N` means.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/runtime.rs [hunk 3]
```diff
@@ -798,7 +798,14 @@ impl CosmosDriverRuntimeBuilder {
             connection_pool,
             bootstrap_transport,
             http_client_factory,
-            env_operation_options: Arc::new(OperationOptions::from_env()),
+            env_operation_options: Arc::new(OperationOptions {
+                // Nested option groups are not populated by the parent's
+                // `from_env`, so the throttling group is loaded explicitly to
+                // keep `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT` honored at the
+                // environment layer.
+                throttling_retry_options: Some(crate::options::ThrottlingRetryOptions::from_env()),
```
Comment (kundadebdatta) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/runtime.rs:814]: **🟡 Recommendation** · Maintainability: Macro Recursion

**Fragile env-loading pattern — future nested option groups will silently miss their env vars**

The `#[option(nested)]` macro generates a `from_env_vars` that does **not** recurse into nested option groups. The author has correctly compensated by manually invoking `ThrottlingRetryOptions::from_env()` here, but this is now the **only** place in the codebase where the contract is enforced. The new test `from_env_vars_parses_known_vars` itself documents the gap:

`
ust
// Nested option groups are not populated by the parent's `from_env`;
// they are loaded separately (see `throttling_retry_options_from_env`).
assert!(options.throttling_retry_options.is_none());
`

**Why this is a latent risk:** any future nested option group added under `OperationOptions` (e.g., a future `ConnectionOptions`, `RetryOptions`, `DiagnosticsOptions`) will silently lose env-layer support unless someone remembers to add a sibling explicit `::from_env()` call at this exact site. There's no compile-time check.

**Fix options (in decreasing preference):**
1. **Best** — teach the `CosmosOptions` derive macro to recursively call `from_env_vars` for `#[option(nested)]` fields and return `Some(...)` when any inner field was populated. This block then simplifies to `..OperationOptions::from_env()` and the test gap disappears.
2. **Defensive** — add a compile-time check or doc test that fails when a new `#[option(nested)]` field appears on `OperationOptions` without a corresponding `::from_env()` call here. Harder to wire up cleanly.
3. **Minimum** — add a `// NOTE: when adding a new nested option group to OperationOptions, you MUST add an explicit ::from_env() call here` comment at this site (the existing comment explains the *why* but not the *invariant*).

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/components.rs [hunk 4]
```diff
@@ -331,21 +331,33 @@ pub(crate) struct ThrottleRetryState {
 }
 
 /// Hard-coded defaults for throttle retry.
-const DEFAULT_MAX_THROTTLE_ATTEMPTS: u32 = 9;
-const DEFAULT_MAX_THROTTLE_WAIT: Duration = Duration::from_secs(30);
+pub(crate) const DEFAULT_MAX_THROTTLE_ATTEMPTS: u32 = 9;
+pub(crate) const DEFAULT_MAX_THROTTLE_WAIT: Duration = Duration::from_secs(30);
```
Comment (kundadebdatta) [inline sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/components.rs]: **🟡 Recommendation** · Open Thread: Defaults

**Resolve the 30s vs 15s default question before merge**

@FabianMeiswinkel raised this on an open review thread: `do we really want to keep 30 seconds - seems a lot - now is the time to consider any changes to the defaults?` — and you replied proposing 15s but haven't committed a change. Now-or-never: lowering the default post-merge is technically a behavior-breaking change for workloads that depended on absorbing up to 30s of throttle backoff.

**If 15s wins:**
1. Update `DEFAULT_MAX_THROTTLE_WAIT` here to `Duration::from_secs(15)`.
2. Update doc comments in `operation_options.rs` (`Default: 15 seconds`) and `cosmos_client_builder.rs` setter doc.
3. Update `ConfigurationOptions.md` table row.
4. Update CHANGELOG entries in both crates.
5. Update PR description.
6. (Worth re-reading) The test `fault_injection_429_honors_configurable_throttle_retry_count` sets the wait to 300s explicitly so the attempt count is the only limiter — so it shouldn't regress with the lower default.

**If 30s wins:** post a reply on the open thread explaining the .NET / Java parity rationale and request thread resolution so the PR can ship without ambiguity.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs [hunk 5]
```diff
@@ -351,10 +418,13 @@ impl CosmosClientBuilder {
...
-            azure_data_cosmos_driver::options::OperationOptionsBuilder::new()
-                .with_per_partition_circuit_breaker_enabled(ppcb_enabled)
-                .build();
+        //
+        // Start from the client-level operation defaults (set via the builder,
+        // e.g. `with_max_retry_attempts_on_throttled_requests`) so they are
+        // forwarded to the driver's runtime layer, then force the resolved
+        // PPCB default on top.
+        let mut runtime_operation_options = self.options.operation.clone();
+        runtime_operation_options.per_partition_circuit_breaker_enabled = Some(ppcb_enabled);
```
Comment (kundadebdatta) [inline sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs]: **🟢 Suggestion** · Maintainability: Latent Footgun (carry-over from prior round)

**`build()` silently clobbers any future SDK-builder-set `per_partition_circuit_breaker_enabled`**

This is the same issue I raised in my Round 2 review (finding #6) — now relevant again because this PR established the pattern `runtime_operation_options = self.options.operation.clone()` (which is the right call) and then unconditionally overrides one specific field. Today it's a no-op since there's no SDK-builder setter for PPCB. The risk is the next contributor who adds one (or wires PPCB through a different SDK-level config knob) won't notice the override and their value will be silently dropped.

```suggestion
        let mut runtime_operation_options = self.options.operation.clone();
        // Force the env-derived PPCB default only when the SDK builder hasn't
        // already set an explicit value (no public SDK setter exists today,
        // but guard against silently clobbering one added in a future PR).
        if runtime_operation_options.per_partition_circuit_breaker_enabled.is_none() {
            runtime_operation_options.per_partition_circuit_breaker_enabled = Some(ppcb_enabled);
        }
```

Not a blocker — carry forward as follow-up if there's release pressure.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs [hunk 6]
```diff
@@ -281,6 +358,21 @@ mod tests {
         assert_eq!(options.max_session_retry_count, Some(3));
         // Fields without env annotation remain None
         assert!(options.excluded_regions.is_none());
+        // Nested option groups are not populated by the parent's `from_env`;
+        // they are loaded separately (see `throttling_retry_options_from_env`).
+        assert!(options.throttling_retry_options.is_none());
+    }
+
+    #[test]
+    fn throttling_retry_options_from_env() {
```
Comment (kundadebdatta) [inline sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs:378]: **🟢 Suggestion** · Maintainability: Test Coverage

**Add a unit test for layered override of the nested `ThrottlingRetryOptions`**

This new test covers env loading well, but there's no companion unit test that verifies the nested struct's **layered resolution** — i.e., when `throttling_retry_options.max_retry_count` is `Some(2)` at the operation layer and `Some(9)` at the runtime layer, that the view resolves to `Some(2)` (operation wins). The integration test `fault_injection_429_honors_configurable_throttle_retry_count` only sets values at the runtime layer.

The `#[option(nested)]` macro's **per-inner-field** layered resolution is the load-bearing contract for the whole feature. A regression in the macro that broke nested per-field resolution would only surface as "per-op throttle overrides are silently ignored" — detectable end-to-end but invisible to the current unit test suite.

**Suggested addition** (inside the same `tests` module):

`
ust
#[test]
fn nested_throttling_retry_options_resolves_across_layers() {
    use std::sync::Arc;
    let runtime = Arc::new(OperationOptions {
        throttling_retry_options: Some(ThrottlingRetryOptions {
            max_retry_count: Some(9),
            max_retry_wait_time: Some(Duration::from_secs(30)),
        }),
        ..Default::default()
    });
    let operation = OperationOptions {
        throttling_retry_options: Some(ThrottlingRetryOptions {
            max_retry_count: Some(0), // operation overrides
            max_retry_wait_time: None, // inherit from runtime
        }),
        ..Default::default()
    };
    let view =
        OperationOptionsView::new(None, Some(runtime), None, Some(&operation));
    let throttling = view.throttling_retry_options();
    assert_eq!(throttling.max_retry_count(), Some(&0),
        "operation override wins");
    assert_eq!(throttling.max_retry_wait_time(), Some(&Duration::from_secs(30)),
        "missing inner field falls through to runtime layer");
}
`$footer

File: sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs [hunk 7]
```diff
@@ -119,6 +119,73 @@ impl CosmosClientBuilder {
...
+    /// throttled requests, surfacing the first 429 to the caller.
+    ///
+    /// **Default**: `9`.
+    ///
+    /// **Scope**: applies *per transport-pipeline invocation*, not per
+    /// logical operation. An operation that fans out across regions
+    /// (failover, hedging) enters the transport pipeline once per leg, each
+    /// with a fresh throttle-retry budget. To cap an operation's *total*
+    /// wall-clock time, configure an end-to-end latency policy on
+    /// [`OperationOptions::end_to_end_latency_policy`](crate::OperationOptions).
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs]: Thanks! Fixed now.

File: sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs [hunk 8]
```diff
@@ -119,6 +119,73 @@ impl CosmosClientBuilder {
...
+    /// `CosmosClientOptions.MaxRetryWaitTimeOnRateLimitedRequests`. Once the
+    /// accumulated retry delay would exceed this budget, no further throttle
+    /// retry is attempted.
+    ///
+    /// **Default**: 30 seconds.
+    ///
+    /// **Scope**: same per-invocation scope as
+    /// [`with_max_retry_attempts_on_throttled_requests`](Self::with_max_retry_attempts_on_throttled_requests)
+    /// — this is *not* a per-operation cap. Configure
+    /// [`OperationOptions::end_to_end_latency_policy`](crate::OperationOptions)
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs]: Thanks! Fixed now.

File: sdk/cosmos/azure_data_cosmos_driver/tests/emulator_tests/driver_fault_injection.rs [hunk 9]
```diff
@@ -331,3 +333,116 @@ pub async fn fault_injection_connection_error() -> Result<(), Box<dyn Error>> {
     })
     .await
 }
+
+/// End-to-end validation that the configurable 429 (throttle) retry budget —
+/// [`OperationOptions::max_throttle_retry_count`](azure_data_cosmos_driver::options::OperationOptions::max_throttle_retry_count)
+/// — is honored across the full driver stack against a live emulator account.
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/tests/emulator_tests/driver_fault_injection.rs]: Thanks! Fixed now.

File: sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs (hunk 2 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/src/options/operation_options.rs:227]: This is now refactored and fixed.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/components.rs (hunk 4 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/pipeline/components.rs]: This is now refactored and fixed.

File: sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs (hunk 5 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/src/clients/cosmos_client_builder.rs]: This is now refactored and fixed.

File: sdk/cosmos/azure_data_cosmos_driver/src/driver/runtime.rs (hunk 3 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/src/driver/runtime.rs:814]: This is now refactored and fixed.

**Review by analogrelay — APPROVED** [review]

**Review by simorenoh — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

File: sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md [hunk 10]
```diff
@@ -4,6 +4,8 @@
...
...`ThrottlingRetryOptions` group (mirroring the .NET and Java SDKs' `ThrottlingRetryOptions`): `max_retry_count` (env `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT`, default `9`, `0` disables retries) caps the number of 429 retries, and `max_retry_wait_time` (default `15s`) caps the cumulative wait across them. Each field resolves independently through the standard operation → account → runtime → env layering and is threaded into the transport-level throttle retry loop via `ThrottleRetryState::with_limits`. The one-shot forced-final-retry safety net in `execute_transport_pipeline` is gated on `attempt_count < max_attempts`, so a configured `max_retry_count = N` produces at most `N + 1` total requests on the wire (1 initial + N retries) — matching the .NET-parity `MaxRetryAttemptsOnRateLimitedRequests` contract. The forced-final retry remains active when the cumulative-wait budget is the limiter rather than the count. Both budgets are scoped per `execute_transport_pipeline` invocation; an operation that fans out across regions (failover, hedging) starts a fresh throttle budget per leg, so per-operation total wall-clock time should be bounded via `OperationOptions::end_to_end_latency_policy`.
```
Comment (tvaron3) [inline sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md]: pr links

File: sdk/cosmos/azure_data_cosmos/CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,8 @@
 
 ### Features Added
 
+- Added configurable retry limits for throttled (HTTP 429, rate-limited) requests, mirroring the .NET and Java SDKs' `ThrottlingRetryOptions`. A new nested `ThrottlingRetryOptions` group on `OperationOptions` (field `throttling_retry_options`) carries `max_retry_count` (env `AZURE_COSMOS_MAX_THROTTLE_RETRY_COUNT`, default `9`, `0` disables throttle retries) and `max_retry_wait_time` (default `15s`), settable per-request via `OperationOptions`/`OperationOptionsBuilder` and `ThrottlingRetryOptionsBuilder`. New client-wide setter `CosmosClientBuilder::with_throttling_retry_options(ThrottlingRetryOptions)` forwards the group as runtime-layer defaults. Both budgets apply *per transport-pipeline invocation*, not per logical operation — an operation that fans out across regions (failover, hedging) starts a fresh budget per leg; use `OperationOptions::end_to_end_latency_policy` to bound total per-operation wall-clock time.
```
Comment (tvaron3) [inline sdk/cosmos/azure_data_cosmos/CHANGELOG.md]: pr links

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md [hunk 12]
```diff
@@ -95,7 +95,7 @@ For single-write accounts, retry cycles through the available endpoint(s). For m
 
 | Substatus | Action | Budget |
 |-----------|--------|--------|
-| — (standard) | Local retry with backoff | 9 attempts / 30s total |
+| — (standard) | Local retry with backoff | 9 attempts / 15s total |
```
Comment (tvaron3) [inline sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md]: curious, what motivated the change to 15 seconds? other sdks use 30 seconds.

**Review by analogrelay — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md (hunk 12 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/ErrorCodesAndRetries.md]: It should not be. It was based off an earlier conversation with Fabian on reducing this, but we agreed to defer that at a later point in time. Eventually we want to reduce the default retry duration to 15 secs but not at this point.

Fixed and updated the documentation at other places.

File: sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md (hunk 10 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/CHANGELOG.md]: Updated.

File: sdk/cosmos/azure_data_cosmos/CHANGELOG.md (hunk 11 above)
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/CHANGELOG.md]: Updated.

**Review by analogrelay — APPROVED** [review]

**Review by simorenoh — APPROVED** [review]

**Review by tvaron3 — APPROVED** [review]
